zio/parquetio/iterator.go

```diff
@@ -0,0 +1,723 @@
+			r.vals = r.vals[:n]
+		} else {
+			r.vals = make([]int64, n, n)
+		}
+
+		mask := makeMask(r.width)
+		var iv uint64
+		havebits := 0
+
+		if len(r.buf) < (groups * r.width) {
+			panic(fmt.Sprintf("need %d bytes for packed but have %d", groups*r.width, len(r.buf)))
+		}
+		bi := 0
+		for i := 0; i < n; i++ {
+			for havebits < r.width {
+				iv |= uint64(r.buf[bi]) << havebits
+				havebits += 8
+				bi++
+			}
+			r.vals[i] = int64(iv & mask)
+			iv >>= r.width
+			havebits -= r.width
+		}
+		r.buf = r.buf[r.width*groups:]
+	}
+}
+
+func (r *hybridReader) peekInt64() int64 {
+	r.fillVals()
+	return r.vals[0]
+}
+
+func (r *hybridReader) nextInt64() int64 {
+	r.fillVals()
+	ret := r.vals[0]
+	r.vals = r.vals[1:]
+	return ret
+}
+
+// nullReader repeatedly returns a default value for some data type.
+// This is a development tool that allows us to "handle" unsupported
+// encodings without panicing or failing.  Eventually this should be
+// removed.
+type nullReader struct {
+}
+
+func (r *nullReader) nextBoolean() bool {
+	return false
+}
+
+func (r *nullReader) nextInt32() int32 {
+	return 0
+}
+
+func (r *nullReader) nextInt64() int64 {
+	return 0
+}
+
+func (r *nullReader) nextFloat() float64 {
+	return 0
+}
+
+func (r *nullReader) nextDouble() float64 {
+	return 0
+}
+
+func (r *nullReader) nextByteArray() []byte {
+	return nil
+}
+
+// plainBooleanReader decodes BOOLEAN-typed values encoded with the
+// parquet PLAIN encoding described at:
+// https://github.com/apache/parquet-format/blob/master/Encodings.md#plain-plain--0
+type plainBooleanReader struct {
+	buf     []byte
+	current byte
+	bits    int
+}
+
+func (r *plainBooleanReader) nextBoolean() bool {
+	if r.bits == 0 {
+		r.current = r.buf[0]
+		r.buf = r.buf[1:]
+		r.bits = 8
+	}
+	b := (r.current & 1) == 1
+	r.current >>= 1
+	r.bits -= 1
+	return b
+}
+
+func (r *plainBooleanReader) nextInt32() int32 {
+	panic("cannot read INT32 from PLAIN BOOLEAN reader")
+}
+
+func (r *plainBooleanReader) nextInt64() int64 {
+	panic("cannot read INT64 from PLAIN BOOLEAN reader")
+}
+
+func (r *plainBooleanReader) nextFloat() float64 {
+	panic("cannot read FLOAT from PLAIN BOOLEAN reader")
+}
+
+func (r *plainBooleanReader) nextDouble() float64 {
+	panic("cannot read DOUBLE from PLAIN BOOLEAN reader")
+}
+
+func (r *plainBooleanReader) nextByteArray() []byte {
+	panic("cannot read BYTE_ARRAY from PLAIN BOOLEAN reader")
+}
+
+// plainReader decodes values encoded with the parquet PLAIN encoding
+// for all types other than BOOLEAN.  PLAIN encoding is described at:
+// https://github.com/apache/parquet-format/blob/master/Encodings.md#plain-plain--0
+// Handle Parquet PLAIN encoding type
+type plainReader struct {
+	buf []byte
+}
+
+func (r *plainReader) nextBoolean() bool {
+	// panic("implement plain boolean reader")
+	return false
+}
+
+func (r *plainReader) nextInt32() int32 {
+	if len(r.buf) < 4 {
+		panic("underflow in PLAIN INT32")
+	}
+
+	ret := binary.LittleEndian.Uint32(r.buf[:4])
```

Comment:
The runtime already does bounds checking, so there's no need to do it yourself in any of these functions.

Reply:
These were placeholders to eventually turn into real error handling that propagates to the caller instead of panics but I can remove them if you prefer.